lib/rules/no-unused-vars.js

```diff
@@ -624,10 +624,26 @@ module.exports = {
 
                     // Report the first declaration.
                     if (unusedVar.defs.length > 0) {
+                        let writeReferences;
+
+                        if (unusedVar.references.length > 0) {
+                            writeReferences = unusedVar.references.filter(ref => ref.isWrite());
+                        }
+
+                        let referenceToReport;
+
+                        if (unusedVar.references.length) {
+
+                            // report last write reference, https://github.com/eslint/eslint/issues/14324
+                            if (Array.isArray(writeReferences) && writeReferences.length) {
+                                referenceToReport = writeReferences[writeReferences.length - 1];
```

Comment:
It could be still confusing if there are references in inner functions, for example:

```js
/* eslint no-unused-vars: error */

let foo;

init();

foo = foo + 2; // should report here?

function init() {
    foo = 1; // reports here
}

```

I think we should check if all write references are in the same function/global/module scope where the variable is defined (`reference.from.variableScope` should be the same as `unusedVar.scope.variableScope`), and if that isn't the case then just fallback to reporting the definition.

Reply:
Fixed 👍🏻

Reply:
I was thinking about reporting the definition (`foo` in `let foo;`) in this case, but it's probably fine to report the last write reference in the same `variableScope` even if there are write references in inner functions, as you implemented now. It _might_ be still confusing only in some edge cases.